File: src/main/java/engineer/carrot/warren/ssl/FingerprintX509TrustManager.java
package engineer.carrot.warren.ssl;

import com.google.common.io.BaseEncoding;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.net.ssl.X509TrustManager;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.cert.CertificateException;
import java.security.cert.X509Certificate;
import java.util.Set;

public class FingerprintX509TrustManager implements X509TrustManager {
    final Logger LOGGER = LoggerFactory.getLogger(FingerprintX509TrustManager.class);

    public Set<String> forciblyAcceptedSHA1Fingerprints;

    public FingerprintX509TrustManager(Set<String> forciblyAcceptedSHA1Fingerprints) {
        this.forciblyAcceptedSHA1Fingerprints = forciblyAcceptedSHA1Fingerprints;
    }

    @Override
    public void checkClientTrusted(X509Certificate[] x509Certificates, String s) throws CertificateException {
        throw new CertificateException("Forcible Trust Manager is not made to verify client certificates");
    }

    @Override
    public void checkServerTrusted(X509Certificate[] x509Certificates, String s) throws CertificateException {
        MessageDigest sha1Digest = null;
        try {
            sha1Digest = MessageDigest.getInstance("SHA1");
        } catch (NoSuchAlgorithmException e) {
            LOGGER.error("Couldn't get SHA1 instance: {}", e);

            throw new CertificateException("Couldn't check server trust because SHA1 digest instance is missing!");
        }

        boolean allTrusted = true;

        LOGGER.info("Checking presented certificates against forcibly trusted: ");
        for (X509Certificate certificate : x509Certificates) {
            certificate.checkValidity();

            sha1Digest.update(certificate.getEncoded());
            String sha1DigestString = BaseEncoding.base16().encode(sha1Digest.digest());
            if (this.forciblyAcceptedSHA1Fingerprints.contains(sha1DigestString)) {
                LOGGER.info(" {} IS trusted", sha1DigestString);
            } else {
                LOGGER.info(" {} IS NOT trusted", sha1DigestString);
                allTrusted = false;
            }
        }

        if (!allTrusted) {
            throw new CertificateException("Certificates were presented with SHA1 signatures that we DO NOT trust!");
        } else {
            LOGGER.info("All presented certificates were forcibly trusted by us");
        }
    }

    @Override
    public X509Certificate[] getAcceptedIssuers() {
        return new X509Certificate[0];
    }
}

File: src/main/kotlin/engineer/carrot/warren/warren/ssl/FingerprintX509TrustManager.kt
package engineer.carrot.warren.warren.ssl

import com.google.common.io.BaseEncoding

import javax.net.ssl.X509TrustManager
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException
import java.security.cert.CertificateException
import java.security.cert.X509Certificate

internal class FingerprintX509TrustManager(private val forciblyAcceptedSHA1Fingerprints: Set<String>) : X509TrustManager {

    @Throws(CertificateException::class)
    override fun checkClientTrusted(x509Certificates: Array<X509Certificate>, s: String) {
        throw CertificateException("Forcible Trust Manager is not made to verify client certificates")
    }

    @Throws(CertificateException::class)
    override fun checkServerTrusted(x509Certificates: Array<X509Certificate>, s: String) {
        var sha1Digest = try {
            MessageDigest.getInstance("SHA1")
        } catch (e: NoSuchAlgorithmException) {
            println("Couldn't get SHA1 instance: $e")

            throw CertificateException("Couldn't check server trust because SHA1 digest instance is missing!")
        }

        var allTrusted = true

        println("Checking presented certificates against forcibly trusted: ")
        for (certificate in x509Certificates) {
            certificate.checkValidity()

            sha1Digest.update(certificate.encoded)
            val sha1DigestString = BaseEncoding.base16().encode(sha1Digest.digest())
            if (this.forciblyAcceptedSHA1Fingerprints.contains(sha1DigestString)) {
                println(" $sha1DigestString IS trusted")
            } else {
                println(" $sha1DigestString IS NOT trusted")
                allTrusted = false
            }
        }

        if (!allTrusted) {
            throw CertificateException("Certificates were presented with SHA1 signatures that we DO NOT trust!")
        } else {
            println("All presented certificates were forcibly trusted by us")
        }
    }

    override fun getAcceptedIssuers(): Array<X509Certificate> {
        return emptyArray()
    }
}
